### ISSUE TITLE
ndbm gdbm problem in make test

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6906]
Migrated from [rt.perl.org#24444](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24444) (status was 'resolved')

Searchable as RT24444$

**Comment by p5pRT** [comment]
### From warrend@mdhost.cse.tek.com

#### Created by warren.dodge@tek.com



I am mailing this on Solaris because the Linux email is not working\.
I cut\-n\-pasted the meat of the Linux bug into this solaris report\.


I just built perl\-5\.8\.1 and had the following error five times during the make
test 

perl&#8203;: relocation error&#8203;: 
\./lib/auto/NDBM\_File/NDBM\_File\.so&#8203;: undefined symbol&#8203;: dbm\_open

In looking at NDBM\_File\.so with ldd I see

&nbsp;        libgdbm\.so\.3 => /proj/wdt/i686\_linux2\.4/lib/libgdbm\.so\.3 \(0x40009000\)
&nbsp;        libc\.so\.6 => /lib/i686/libc\.so\.6 \(0x42000000\)
&nbsp;        /lib/ld\-linux\.so\.2 => /lib/ld\-linux\.so\.2 \(0x80000000\)

Notice it is using the libgdbm \(gnu\) library\. This doesn't seem right\. There
is not a dbm\_open function in that lib file\.

Ther is a /usr/lib/libndbm\.so file on Linux\. I have installed gdbm as seen
above\. 

I found this in the following file ext/NDBM\_File/hints/linux\.pl

\# Some distributions have both gdbm and ndbm
\# Prefer gdbm to avoid the broken ndbm in some distributions
\# \(no null key support\)
\# Jonathan Stowe \<gellyfish@&#8203;gellyfish\.com>
use Config;
$self\->\{LIBS\} = \['\-lgdbm'\] if $Config\{libs\} =~ /\(?&#8203;:^|\\s\)\-lgdbm\(?&#8203;:\\s|$\)/;

Perhaps there is a logic problem in here somewhere??

This is the first time I have built perl on Linux so I can't tell you if it
has worked previously\.

Let me know what information I can provide or what I can do to help solve
this problem\. 

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.1:

Configured by warrend at Fri Nov  7 12:00:08 PST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.20-20.7bigmem, archname=i686-linux-thread-multi
    uname='linux zephyr 2.4.20-20.7bigmem #1 smp mon aug 18 14:34:37 edt 2003 i686 unknown unknown gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -I/proj/wdt/i686_linux2.4/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -I/proj/wdt/i686_linux2.4/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =''
    libpth=/lib /usr/lib /proj/wdt/i686_linux2.4/lib
    libs=-lnsl -lndbm -lgdbm -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.2.5.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/proj/wdt/gnu_i686_linux2.4/perl-5.8.1/lib/5.8.1/i686-linux-thread-multi/CORE'
    cccdlflags='-fpic', lddlflags='-shared'

Locally applied patches:
    


@INC for perl v5.8.1:
    /proj/wdt/gnu_i686_linux2.4/perl-5.8.1/lib/5.8.1/i686-linux-thread-multi
    /proj/wdt/gnu_i686_linux2.4/perl-5.8.1/lib/5.8.1
    /proj/wdt/gnu_i686_linux2.4/perl-5.8.1/addons/lib/site_perl/5.8.1/i686-linux-thread-multi
    /proj/wdt/gnu_i686_linux2.4/perl-5.8.1/addons/lib/site_perl/5.8.1
    /proj/wdt/gnu_i686_linux2.4/perl-5.8.1/addons/lib/site_perl
    .


Environment for perl v5.8.1:
    HOME=/login/warrend/lxhome
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=.:/proj/wdt/bin:/proj/wdt/i686_linux2.4/bin:/usr/X11R6/bin:/login/warrend:/login/warrend/bin:/login/warrend/bin/i686_linux2.4:/bin:/usr/bin:/etc:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Warren L Dodge \(via RT\) wrote&#8203;:
> 
> I just built perl\-5\.8\.1 and had the following error five times during the make
> test 
> 
> perl&#8203;: relocation error&#8203;: 
> \./lib/auto/NDBM\_File/NDBM\_File\.so&#8203;: undefined symbol&#8203;: dbm\_open
> 
> In looking at NDBM\_File\.so with ldd I see
> 
>         libgdbm\.so\.3 => /proj/wdt/i686\_linux2\.4/lib/libgdbm\.so\.3 \(0x40009000\)
>         libc\.so\.6 => /lib/i686/libc\.so\.6 \(0x42000000\)
>         /lib/ld\-linux\.so\.2 => /lib/ld\-linux\.so\.2 \(0x80000000\)
> 
> Notice it is using the libgdbm \(gnu\) library\. This doesn't seem right\. There
> is not a dbm\_open function in that lib file\.
> 
> Ther is a /usr/lib/libndbm\.so file on Linux\. I have installed gdbm as seen
> above\. 

You have two options :
\- don't build NDBM\_File\, by passing the \-Ui\_ndbm command line flag to Configure
\- try to delete the linux NDBM\_File hint file and see if it can figure out
&nbsp;  to link NDBM\_File with libndbm \-\- and if it works correctly\.

It depends on your need for NDBM\_File\.

> I found this in the following file ext/NDBM\_File/hints/linux\.pl
> 
> \# Some distributions have both gdbm and ndbm
> \# Prefer gdbm to avoid the broken ndbm in some distributions
> \# \(no null key support\)
> \# Jonathan Stowe \<gellyfish@&#8203;gellyfish\.com>
> use Config;
> $self\->\{LIBS\} = \['\-lgdbm'\] if $Config\{libs\} =~ /\(?&#8203;:^|\\s\)\-lgdbm\(?&#8203;:\\s|$\)/;
> 
> Perhaps there is a logic problem in here somewhere??

API compatibility between \*dbm libraries is left as an exercise to the
reader :\) I'm not familiar enough with these to comment\.

**Comment by p5pRT** [comment]
### From warren.dodge@tektronix.com


Ok\, I tried two things in this ndbm/gdbm area\.

I turned off the ext/NDBM\_File/hints/linux\.pl file\. I did a new configure and
build\. This did change the test results but there were still errors\. It fixed
the ndbm using the gdbm library file\.

So the new errors are like this one\.

lib/AnyDBM\_File\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.ndbm store returned \-1\, errno 22\, key ""
at \.\./lib/AnyDBM\_File\.t line 111\. 


I also did a new configure and removed ndbm info from the list of libs and
from the loaded extensions\. This made and tested with no errors by skipping
all the tests that had to do with ndbm\.

So\, I don't know what to do/think\. I would think\, if each of the libraries
are available\, that perl should sort things out and "just work"\. When
hundreds of people use a build\, it is very difficult to know what they are
doing\.

Also\, this all works fine on Solaris 8 using all the default choices in the
configure\. 


>  Date&#8203;: 8 Nov 2003 10&#8203;:06&#8203;:29 \-0000
>  From&#8203;: Rafael Garcia\-Suarez \(via RT\) \<perlbug\-followup@&#8203;perl\.org>
>  Reply\-To&#8203;: perlbug\-followup@&#8203;perl\.org
>  X\-RT\-Loop\-Prevention&#8203;: perl
>  RT\-Ticket&#8203;: perl \#24444
>  Managed\-by&#8203;: RT 2\.0\.14 \(http&#8203;://bestpractical\.com/rt/\)
>  RT\-Originator&#8203;: rgarciasuarez@&#8203;free\.fr
>  Content\-Disposition&#8203;: inline
>  
>  Warren L Dodge \(via RT\) wrote&#8203;:
>  > 
>  > I just built perl\-5\.8\.1 and had the following error five times during the make
>  > test 
>  > 
>  > perl&#8203;: relocation error&#8203;: 
>  > \./lib/auto/NDBM\_File/NDBM\_File\.so&#8203;: undefined symbol&#8203;: dbm\_open
>  > 
>  > In looking at NDBM\_File\.so with ldd I see
>  > 
>  >         libgdbm\.so\.3 => /proj/wdt/i686\_linux2\.4/lib/libgdbm\.so\.3 \(0x40009000\)
>  >         libc\.so\.6 => /lib/i686/libc\.so\.6 \(0x42000000\)
>  >         /lib/ld\-linux\.so\.2 => /lib/ld\-linux\.so\.2 \(0x80000000\)
>  > 
>  > Notice it is using the libgdbm \(gnu\) library\. This doesn't seem right\. There
>  > is not a dbm\_open function in that lib file\.
>  > 
>  > Ther is a /usr/lib/libndbm\.so file on Linux\. I have installed gdbm as seen
>  > above\. 
>  
>  You have two options :
>  \- don't build NDBM\_File\, by passing the \-Ui\_ndbm command line flag to Configure
>  \- try to delete the linux NDBM\_File hint file and see if it can figure out
>    to link NDBM\_File with libndbm \-\- and if it works correctly\.
>  
>  It depends on your need for NDBM\_File\.
>  
>  > I found this in the following file ext/NDBM\_File/hints/linux\.pl
>  > 
>  > \# Some distributions have both gdbm and ndbm
>  > \# Prefer gdbm to avoid the broken ndbm in some distributions
>  > \# \(no null key support\)
>  > \# Jonathan Stowe \<gellyfish@&#8203;gellyfish\.com>
>  > use Config;
>  > $self\->\{LIBS\} = \['\-lgdbm'\] if $Config\{libs\} =~ /\(?&#8203;:^|\\s\)\-lgdbm\(?&#8203;:\\s|$\)/;
>  > 
>  > Perhaps there is a logic problem in here somewhere??
>  
>  API compatibility between \*dbm libraries is left as an exercise to the
>  reader :\) I'm not familiar enough with these to comment\.
>  
>

**Comment by p5pRT** [comment]
### From @tux

On Tue 11 Nov 2003 21&#8203;:41\, Warren L Dodge \<warrend@&#8203;mdhost\.cse\.tek\.com> wrote&#8203;:
> 
> Ok\, I tried two things in this ndbm/gdbm area\.

\<bells ringin'>

Do you get anything like this&#8203;:

\# nm /usr/lib/libndbm\.a
nm&#8203;: libndbm\.a&#8203;: File format not recognized

Remove /usr/lib/libndbm\.a and /usr/lib/libndbm\.so\, both likely to be of size
182/184\. They are just \(invalid\) placeholders

> I turned off the ext/NDBM\_File/hints/linux\.pl file\. I did a new configure and
> build\. This did change the test results but there were still errors\. It fixed
> the ndbm using the gdbm library file\.
> 
> So the new errors are like this one\.
> 
> lib/AnyDBM\_File\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.ndbm store returned \-1\, errno 22\, key ""
> at \.\./lib/AnyDBM\_File\.t line 111\. 
> 
> 
> I also did a new configure and removed ndbm info from the list of libs and
> from the loaded extensions\. This made and tested with no errors by skipping
> all the tests that had to do with ndbm\.
> 
> So\, I don't know what to do/think\. I would think\, if each of the libraries
> are available\, that perl should sort things out and "just work"\. When
> hundreds of people use a build\, it is very difficult to know what they are
> doing\.
> 
> Also\, this all works fine on Solaris 8 using all the default choices in the
> configure\. 
> 
> 
> >  Date&#8203;: 8 Nov 2003 10&#8203;:06&#8203;:29 \-0000
> >  From&#8203;: Rafael Garcia\-Suarez \(via RT\) \<perlbug\-followup@&#8203;perl\.org>
> >  Reply\-To&#8203;: perlbug\-followup@&#8203;perl\.org
> >  X\-RT\-Loop\-Prevention&#8203;: perl
> >  RT\-Ticket&#8203;: perl \#24444
> >  Managed\-by&#8203;: RT 2\.0\.14 \(http&#8203;://bestpractical\.com/rt/\)
> >  RT\-Originator&#8203;: rgarciasuarez@&#8203;free\.fr
> >  Content\-Disposition&#8203;: inline
> >  
> >  Warren L Dodge \(via RT\) wrote&#8203;:
> >  > 
> >  > I just built perl\-5\.8\.1 and had the following error five times during the make
> >  > test 
> >  > 
> >  > perl&#8203;: relocation error&#8203;: 
> >  > \./lib/auto/NDBM\_File/NDBM\_File\.so&#8203;: undefined symbol&#8203;: dbm\_open
> >  > 
> >  > In looking at NDBM\_File\.so with ldd I see
> >  > 
> >  >         libgdbm\.so\.3 => /proj/wdt/i686\_linux2\.4/lib/libgdbm\.so\.3 \(0x40009000\)
> >  >         libc\.so\.6 => /lib/i686/libc\.so\.6 \(0x42000000\)
> >  >         /lib/ld\-linux\.so\.2 => /lib/ld\-linux\.so\.2 \(0x80000000\)
> >  > 
> >  > Notice it is using the libgdbm \(gnu\) library\. This doesn't seem right\. There
> >  > is not a dbm\_open function in that lib file\.
> >  > 
> >  > Ther is a /usr/lib/libndbm\.so file on Linux\. I have installed gdbm as seen
> >  > above\. 
> >  
> >  You have two options :
> >  \- don't build NDBM\_File\, by passing the \-Ui\_ndbm command line flag to Configure
> >  \- try to delete the linux NDBM\_File hint file and see if it can figure out
> >    to link NDBM\_File with libndbm \-\- and if it works correctly\.
> >  
> >  It depends on your need for NDBM\_File\.
> >  
> >  > I found this in the following file ext/NDBM\_File/hints/linux\.pl
> >  > 
> >  > \# Some distributions have both gdbm and ndbm
> >  > \# Prefer gdbm to avoid the broken ndbm in some distributions
> >  > \# \(no null key support\)
> >  > \# Jonathan Stowe \<gellyfish@&#8203;gellyfish\.com>
> >  > use Config;
> >  > $self\->\{LIBS\} = \['\-lgdbm'\] if $Config\{libs\} =~ /\(?&#8203;:^|\\s\)\-lgdbm\(?&#8203;:\\s|$\)/;
> >  > 
> >  > Perhaps there is a logic problem in here somewhere??
> >  
> >  API compatibility between \*dbm libraries is left as an exercise to the
> >  reader :\) I'm not familiar enough with these to comment\.
> >  
> >  

\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using perl\-5\.6\.1\, 5\.8\.0\, & 5\.9\.x\, and 806 on  HP\-UX 10\.20 & 11\.00\, 11i\,
&nbsp;   AIX 4\.3\, SuSE 8\.2\, and Win2k\.           http&#8203;://www\.cmve\.net/~merijn/
http&#8203;://archives\.develooper\.com/daily\-build@&#8203;perl\.org/   perl\-qa@&#8203;perl\.org
send smoke reports to&#8203;: smokers\-reports@&#8203;perl\.org\, QA&#8203;: http&#8203;://qa\.perl\.org

**Comment by p5pRT** [comment]
### From warrend@mdhost.cse.tek.com



ll  /usr/lib/libndb\*
lrwxrwxrwx    1 root     root      8 Aug 19 08&#8203;:02 /usr/lib/libndbm\.a \-> libdb2\.a
lrwxrwxrwx    1 root     root     11 Aug 19 08&#8203;:02 /usr/lib/libndbm\.so \-> libdb2\.so\.3

ll \-L  /usr/lib/libndb\*
\-rw\-r\-\-r\-\-    1 root     root 392034 Apr  2  2002 /usr/lib/libndbm\.a
\-rwxr\-xr\-x    1 root     root 277324 Apr  2  2002 /usr/lib/libndbm\.so\*

Each of these do the nm just fine\.

The system I am using is a "standard" linux distribution\. It is also a
backroom machine where I can't do anything as root\.


Here is the output of one test run

\./perl  lib/AnyDBM\_File\.t
1\.\.12
ok 1 \- Tie
ok 2 \- File permissions
ok 3 \- Hash created empty
ok 4 \- Re\-tie hash
ok 5 \- $\\\#keys == $\\\#values
ok 6 \- keys and values match
ok 7 \- Correct number of keys
ndbm store returned \-1\, errno 22\, key "" at lib/AnyDBM\_File\.t line 111\.

This is doing  $h\{''\} = 'bar';

I can't figure out how to trace this error down\. 

I really don't understand all the "dbm" formats or how to use them so I am
not much help in this area\. 

In the past on solaris I just ran the default Configure and everything
worked\. On Linux is where the default doesn't pass the tests\.

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Nov 12\, 2003 at 09&#8203;:28&#8203;:57AM \-0800\, Warren L Dodge wrote&#8203;:

> ndbm store returned \-1\, errno 22\, key "" at lib/AnyDBM\_File\.t line 111\.
> 
> This is doing  $h\{''\} = 'bar';
> 
> I can't figure out how to trace this error down\. 

Was there ever a version of ndbm that didn't allow empty keys?
I know that one or other DBM implementation didn't \(for a while\, until
they were convinced that this was a silly thing to disallow\)

Nicholas Clark

**Comment by p5pRT** [comment]
### From @schwern

On Wed\, Nov 12\, 2003 at 09&#8203;:28&#8203;:57AM \-0800\, Warren L Dodge wrote&#8203;:
> The system I am using is a "standard" linux distribution\. It is also a
> backroom machine where I can't do anything as root\.

The great thing about standard Linux distributions is there's so many to 
choose from\! :\)

Knowing which Linux distribution you're using would help\.


\-\- 
Michael G Schwern        schwern@&#8203;pobox\.com  http&#8203;://www\.pobox\.com/~schwern/
Remember\, any tool can be the right tool\.
&nbsp;    \-\- Red Green

**Comment by p5pRT** [comment]
### From warrend@mdhost.cse.tek.com



If you are asking me about this\, I have no idea\.

>  Date&#8203;: 12 Nov 2003 21&#8203;:46&#8203;:39 \-0000
>  From&#8203;: Nicholas Clark \(via RT\) \<perlbug\-followup@&#8203;perl\.org>
>  Reply\-To&#8203;: perlbug\-followup@&#8203;perl\.org
>  X\-RT\-Loop\-Prevention&#8203;: perl
>  RT\-Ticket&#8203;: perl \#24444
>  Managed\-by&#8203;: RT 2\.0\.14 \(http&#8203;://bestpractical\.com/rt/\)
>  RT\-Originator&#8203;: nick@&#8203;ccl4\.org
>  Content\-Disposition&#8203;: inline
>  
>  On Wed\, Nov 12\, 2003 at 09&#8203;:28&#8203;:57AM \-0800\, Warren L Dodge wrote&#8203;:
>  
>  > ndbm store returned \-1\, errno 22\, key "" at lib/AnyDBM\_File\.t line 111\.
>  > 
>  > This is doing  $h\{''\} = 'bar';
>  > 
>  > I can't figure out how to trace this error down\. 
>  
>  Was there ever a version of ndbm that didn't allow empty keys?
>  I know that one or other DBM implementation didn't \(for a while\, until
>  they were convinced that this was a silly thing to disallow\)
>  
>  Nicholas Clark
>  
>

**Comment by p5pRT** [comment]
### From swtaylor@gentoo.org

gdbm\-1\.8\.3 moved the "old" ndbm functions from libgdbm\.so to
libgdbm\_compat\.so \- I've had good luck making the following change&#8203;:



\-\-\- perl\-5\.8\.2/ext/NDBM\_File/hints/linux\.pl\.orig    2003\-11\-11
00&#8203;:50&#8203;:43\.924885744 \-0700
\+\+\+ perl\-5\.8\.2/ext/NDBM\_File/hints/linux\.pl    2003\-11\-11
00&#8203;:51&#8203;:05\.631585824 \-0700
@&#8203;@&#8203; \-3\,4 \+3\,4 @&#8203;@&#8203;
 \# \(no null key support\)
 \# Jonathan Stowe \<gellyfish@&#8203;gellyfish\.com>
 use Config;
\-$self\->\{LIBS\} = \['\-lgdbm'\] if $Config\{libs\} =~ /\(?&#8203;:^|\\s\)\-lgdbm\(?&#8203;:\\s|$\)/;
\+$self\->\{LIBS\} = \['\-lgdbm \-lgdbm\_compat'\] if $Config\{libs\} =~
/\(?&#8203;:^|\\s\)\-lgdbm\(?&#8203;:\\s|$\)/;

**Comment by p5pRT** [comment]
### From warren.dodge@tektronix.com



This seemed to work fine\. I also had to adjust the ODBM hints file to make
those tests pass\.

What I found was that the Redhat release we are using does not provide the
libgdbm\-compat library\. Or at least we could not find it\.

I built gdbm and installed the compat lib to make it work\.

Perhaps this is an omission in the Linux release?


uname \-a
Linux tekcs22 2\.4\.21\-4\.ELhugemem \#1 SMP Fri Oct 3 17&#8203;:31&#8203;:20 EDT 2003 i686 i686 i386 GNU/Linux
with libc\-2\.3\.2\.so

>  Date&#8203;: 21 Jan 2004 22&#8203;:33&#8203;:27 \-0000
>  From&#8203;: "Scott Taylor via RT" \<perlbug\-followup@&#8203;perl\.org>
>  Reply\-To&#8203;: perlbug\-followup@&#8203;perl\.org
>  X\-RT\-Loop\-Prevention&#8203;: perl
>  RT\-Ticket&#8203;: perl \#24444
>  Managed\-by&#8203;: RT 3\.0\.8 \(http&#8203;://www\.bestpractical\.com/rt/\)
>  RT\-Originator&#8203;: scott@&#8203;303underground\.com
>  
>  gdbm\-1\.8\.3 moved the "old" ndbm functions from libgdbm\.so to
>  libgdbm\_compat\.so \- I've had good luck making the following change&#8203;:
>  
>  
>  
>  \-\-\- perl\-5\.8\.2/ext/NDBM\_File/hints/linux\.pl\.orig    2003\-11\-11
>  00&#8203;:50&#8203;:43\.924885744 \-0700
>  \+\+\+ perl\-5\.8\.2/ext/NDBM\_File/hints/linux\.pl    2003\-11\-11
>  00&#8203;:51&#8203;:05\.631585824 \-0700
>  @&#8203;@&#8203; \-3\,4 \+3\,4 @&#8203;@&#8203;
>   \# \(no null key support\)
>   \# Jonathan Stowe \<gellyfish@&#8203;gellyfish\.com>
>   use Config;
>  \-$self\->\{LIBS\} = \['\-lgdbm'\] if $Config\{libs\} =~ /\(?&#8203;:^|\\s\)\-lgdbm\(?&#8203;:\\s|$\)/;
>  \+$self\->\{LIBS\} = \['\-lgdbm \-lgdbm\_compat'\] if $Config\{libs\} =~ /\(?&#8203;:^|\\s\)\-lgdbm\(?&#8203;:\\s|$\)/;
>  
>

**Comment by p5pRT** [comment]
### From swtaylor@gentoo.org

I ran into this issue when gdbm\-1\.8\.3 was added to the unstable tree of
gentoo\. It was rather quickly just masked and moved to the back burner\.
gdbm\-1\.8\.2 still had all the ndbm calls in its main library\. A number of
things seem to still be using things moved into the gdbm\_compat
\(spamassassin's bayes database was the most obvious one i was hit by\)


Blue root \# uname \-a
Linux Blue 2\.6\.1\-gentoo \#2 Mon Jan 12 16&#8203;:07&#8203;:11 MST 2004 i686 AMD
Athlon\(tm\) XP 1500\+ AuthenticAMD GNU/Linux
Blue root \# emerge \-p gdbm perl

These are the packages that I would merge\, in order&#8203;:

Calculating dependencies \.\.\.done\!
\[ebuild   R   \] sys\-libs/gdbm\-1\.8\.3  
\[ebuild   R   \] dev\-lang/perl\-5\.8\.3\-r1  



> \[warren\.dodge@&#8203;tektronix\.com \- Thu Jan 22 15&#8203;:37&#8203;:27 2004\]&#8203;:
> This seemed to work fine\. I also had to adjust the ODBM hints file to
> make those tests pass\.
> 
> What I found was that the Redhat release we are using does not provide
> the libgdbm\-compat library\. Or at least we could not find it\.
> 
> I built gdbm and installed the compat lib to make it work\.
> 
> Perhaps this is an omission in the Linux release?

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @smpeters

On Thu Jan 22 16&#8203;:16&#8203;:05 2004\, swtaylor wrote&#8203;:
> I ran into this issue when gdbm\-1\.8\.3 was added to the unstable tree of
> gentoo\. It was rather quickly just masked and moved to the back burner\.
> gdbm\-1\.8\.2 still had all the ndbm calls in its main library\. A number of
> things seem to still be using things moved into the gdbm\_compat
> \(spamassassin's bayes database was the most obvious one i was hit by\)
> 
> 
> Blue root \# uname \-a
> Linux Blue 2\.6\.1\-gentoo \#2 Mon Jan 12 16&#8203;:07&#8203;:11 MST 2004 i686 AMD
> Athlon\(tm\) XP 1500\+ AuthenticAMD GNU/Linux
> Blue root \# emerge \-p gdbm perl
> 
> These are the packages that I would merge\, in order&#8203;:
> 
> Calculating dependencies \.\.\.done\!
> \[ebuild   R   \] sys\-libs/gdbm\-1\.8\.3  
> \[ebuild   R   \] dev\-lang/perl\-5\.8\.3\-r1  
> 
> 
> 
> > \[warren\.dodge@&#8203;tektronix\.com \- Thu Jan 22 15&#8203;:37&#8203;:27 2004\]&#8203;:
> > This seemed to work fine\. I also had to adjust the ODBM hints file to
> > make those tests pass\.
> > 
> > What I found was that the Redhat release we are using does not provide
> > the libgdbm\-compat library\. Or at least we could not find it\.
> > 
> > I built gdbm and installed the compat lib to make it work\.
> > 
> > Perhaps this is an omission in the Linux release?
> 

This was resolved with change \#22892\.

Steve
